### PR TITLE
Storage: Basic Request Validation

### PR DESCRIPTION
**Pull request by jaschrep-msft**
Introduced a pipeline policy that applies general-purpose assertions to
responses. Can add these configurations individually to, or to all of,
the blob, file, and queue packages.

Added a validation to check if the x-ms-client-request-id header is echoed correctly, should it be present in the response.

### CODE REVIEW

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/ResponseValidationPolicyBuilder.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+     * Fluently applies an optional validation to this policy where, if the response contains the given header, asserts
+     * its value is an echo of the value provided in the request.
+     *
+     * @param headerName The header to validate.
+     * @return This policy.
+     */
+    public ResponseValidationPolicyBuilder optionalEcho(String headerName) {
+        assertions.add((httpResponse, logger) -> {
+            String responseHeaderValue = httpResponse.headerValue(headerName);
+            if (responseHeaderValue != null && !responseHeaderValue.equals(httpResponse.request().headers().value(headerName))) {
```
Comment (gapra-msft) [inline sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/ResponseValidationPolicyBuilder.java]: Does httpResponse.request().headers().value(headerName) throw if headerName is invalid?

Comment (jaschrep-msft) [reply sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/ResponseValidationPolicyBuilder.java]: Javadocs say it returns null.

**Comment by jaschrep-msft** [comment]
Addresses issue #4511

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/ResponseValidationPolicyBuilder.java (hunk 1 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/ResponseValidationPolicyBuilder.java]: Should be fine if it returns null

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/ResponseValidationPolicyBuilder.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    /**
+     * Fluently applies an optional validation to this policy where, if the response contains the given header, asserts
+     * its value is an echo of the value provided in the request.
+     *
+     * @param headerName The header to validate.
+     * @return This policy.
+     */
+    public ResponseValidationPolicyBuilder optionalEcho(String headerName) {
```
[the remark this answers is not in the document]
Comment (jaschrep-msft) [reply sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/ResponseValidationPolicyBuilder.java]: That matches the client builders as well. Will fix.

[the remark this answers is not in the document]
Comment (jaschrep-msft) [reply sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/ResponseValidationPolicyBuilder.java]: Oops, it looks like we log a blank RuntimeException. I can have it construct a message based on the header name.

As for leaving a log message in general, I'd say not on this API, as it's very obviously doing one and only one check, and can easily make a log message as to what was expected and not gotten. In the future, if we need to pass in arbitrary consumers to do more intricate assertions, those APIs can take in log messages. Or the consumer can have the log message on its own.

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/ResponseValidationPolicyBuilder.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        assertions.add((httpResponse, logger) -> {
+            String responseHeaderValue = httpResponse.headerValue(headerName);
+            if (responseHeaderValue != null && !responseHeaderValue.equals(httpResponse.request().headers().value(headerName))) {
+                throw logger.logExceptionAsError(new RuntimeException());
```
Comment (sima-zhu) [inline sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/ResponseValidationPolicyBuilder.java]: Do you need to log specific error msg here? How customers checkout what happened if no error log out.

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/BaseClientBuilder.java [hunk 4]
```diff
@@ -98,6 +102,29 @@ protected HttpPipeline buildPipeline() {
...
+        return builder.build();
+    }
+
+    /**
+     * Applies validation of general-purpose requests to builder. Method-specific validations should be performed in the
+     * convenience layer.
+     *
+     * @param builder Builder to assemble assertions together.
+     */
+    protected abstract void applyServiceSpecificValidations(ResponseValidationPolicyBuilder builder);
```
Comment (sima-zhu) [inline sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/BaseClientBuilder.java]: I did not see any implementation on each builder. Is this method a placeholder or leave it here for some other purpose?

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/ResponseValidationPolicyBuilder.java (hunk 3 above)
Comment (jaschrep-msft) [reply sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/ResponseValidationPolicyBuilder.java]: Mentioned in one of Alan's comments that I missed this. Will be addressed.

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/BaseClientBuilder.java (hunk 4 above)
Comment (jaschrep-msft) [reply sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/BaseClientBuilder.java]: This this is overridden with `final` in BaseBlobClientBuilder, BaseFileClientBuilder, and BaseQueueClientBuilder. They are currently all three no-ops, but this is laying the groundwork for CPK to validate that the key-hash returned is the same as the one sent over the wire. CPK is a blob-only feature, so there was no need to add extra steps to file and queue's pipelines.

**Pull request merged**